src/Shared/Microsoft.DotNet.Internal.DependencyInjection.Testing/DependencyInjectionValidation.cs

```diff
@@ -0,0 +1,154 @@
+
+            var services = new ServiceCollection();
+            if (includeServiceHost)
+            {
+                Environment.SetEnvironmentVariable("ENVIRONMENT", "XUNIT");
+                ServiceHost.ConfigureDefaultServices(services);
+            }
+
+            register(services);
+
+            foreach (ServiceDescriptor service in services)
+            {
+                if (service.ImplementationType == null)
+                {
+                    continue;
+                }
+
+                if (IsExemptType(service.ImplementationType) || IsExemptType(service.ServiceType))
+                {
+                    continue;
+                }
+
+                bool foundConstructable = false;
+                ConstructorInfo[] constructors = service.ImplementationType
+                    .GetConstructors(BindingFlags.Public | BindingFlags.Instance)
+                    .OrderBy(c => c.GetParameters().Length)
+                    .ToArray();
+
+                if (constructors.Length == 0)
+                {
+                    // zero constructor things are implicitly constructable
+                    continue;
+                }
+
+                foreach (ConstructorInfo ctor in constructors)
+                {
+                    bool ctorWorks = true;
+
+                    foreach (ParameterInfo p in ctor.GetParameters())
+                    {
+                        ServiceDescriptor parameterService = services.FirstOrDefault(s => IsMatchingServiceRegistration(s.ServiceType, p.ParameterType));
+                        if (parameterService != null)
+                        {
+                            continue;
+                        }
+
+                        // Save the first error message, since it's likely to be the most useful
+                        if (errorMessage == null)
+                        {
+                            errorMessage = $"Type {service.ImplementationType.FullName} could not find registered definition for parameter {p.Name} of type {p.ParameterType}";
+                        }
+
+                        ctorWorks = false;
+                        break;
+                    }
+
+                    if (ctorWorks)
+                    {
+                        foundConstructable = true;
+                        break;
+                    }
+                }
+
+
+                if (!foundConstructable)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private static bool IsMatchingServiceRegistration(Type serviceType, Type parameterType)
+        {
+            // If it's options, lets make sure they are configured
+            if (parameterType.IsConstructedGenericType)
+            {
+                Type parameterRoot = parameterType.GetGenericTypeDefinition();
+                if (parameterRoot == typeof(IOptions<>) ||
+                    parameterRoot == typeof(IOptionsMonitor<>))
```

Comment:
This also needs IOptionsSnapshot<> I think.

Reply:
Ah, yes.  I just didn't run into of those yet.